building/build-acis/flannel-monitor/src/flannel-monitor/main/flannel-monitor.go

```diff
@@ -0,0 +1,229 @@
+package main
+
+import (
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"log"
+	"os"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"fmt"
+	"net/http"
+	"time"
+	"io/ioutil"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	v12 "k8s.io/api/core/v1"
+	"net"
+)
+
+var (
+	registry = prometheus.NewRegistry()
+
+	talkCheck = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: "flannel",
+		Name: "talk_check",
+		Help: "Check for whether flannel supports container communication",
+	}, []string {"ping_from_host", "ping_to_host"})
+
+	monRecency = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: "flannel",
+		Name: "monitor_recency",
+		Help: "Timestamp for the oldest currently reported metric",
+	}, []string {"ping_from_host"})
+
+	talkTiming = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Namespace: "flannel",
+		Name: "talk_timing",
+		Help: "Timing for flannel communication",
+		Buckets: []float64 {0.1, 0.2, 0.5, 1, 2, 5, 10},
+	}, []string {"ping_from_host", "ping_to_host"})
+
+	monCheck = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: "flannel",
+		Name: "monitor_check",
+		Help: "Whether flannel was able to be monitored from this host (and, therefore, the other metrics are valid)",
+	}, []string {"ping_from_host"})
+
+	dupCheck = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: "flannel",
+		Name: "duplicate_check",
+		Help: "Whether flannel monitoring successfully found no duplicate reflectors",
+	}, []string {"ping_from_host"})
+)
+
+func cycle(clientset *kubernetes.Clientset, namespace string, source_hostip string, client *http.Client) {
+	valid := monCheck.With(prometheus.Labels{"ping_from_host": source_hostip})
+	dupOkay := dupCheck.With(prometheus.Labels{"ping_from_host": source_hostip})
+	monRecencySpec := monRecency.With(prometheus.Labels{"ping_from_host": source_hostip})
+	nodelist, err := clientset.CoreV1().Nodes().List(v1.ListOptions{})
+	if err != nil {
+		talkCheck.Reset()
+		talkTiming.Reset()
+		valid.Set(0)
+		dupOkay.Set(0)
+		log.Printf("could not fetch list of nodes")
+		return
+	}
+	podlist, err := clientset.CoreV1().Pods(namespace).List(v1.ListOptions{LabelSelector: "app=flannel-monitor-reflector"})
+	if err != nil {
+		talkCheck.Reset()
+		talkTiming.Reset()
+		valid.Set(0)
+		dupOkay.Set(0)
+		log.Printf("could not fetch list of flannel-monitor-reflector apps")
+		return
+	}
+
+	check_map := map[string]prometheus.Gauge{}
+	not_found := map[string]struct{}{}
+	timing_map := map[string]prometheus.Observer{}
+	for _, item := range nodelist.Items {
```

Comment:
This code is similar to code in `flannel-monitor-collector.go` which also finds all the nodes' IP addresses, checks all the pods, etc.  Can the similarities be factored out?

Reply:
Not all of them, but I should be able to refactor some of them out.